doc/dashboard: update documentation with accessibility development pr… (#48002)
ceph/ceph: doc/dev/developer_guide/dash-devel.rst

Default notification should have `text-info` color. Success notification should
have `text-success` color. Failure notification should have `text-danger` color.

> Accessibility

Reviewer: I would add this right after chapter `i18n` (and would welcome a name change to `Internationalization`), as both concerns pursue a similar goal: to reach a broader user base by lowering barriers to entry (whether cultural or cognitive).

Reviewer: In any case, I also understand the rationale for putting this under the "UI Style guide". We could add a link there to an outer accessibility section. I was just trying to give "accessibility" the visibility it deserves, and nesting it inside the "UI Style guide" diminished its imporance (IMHO).

Author: Thanks @epuertat

.............

> We follow the `W3C WCAG2.1 <https://www.w3.org/TR/WCAG21/>`_  level A accessibility conformance guidelines. 

Reviewer 2: ```suggestion
Many parts of the Ceph Dashboard are modeled on `Web Content Accessibility Guidelines (WCAG) 2.1 <https://www.w3.org/TR/WCAG21/>`_  level A accessibility conformance guidelines. 
```

> By implementing accessibility best practices, you are also improving the usability of the site for all users.

Reviewer 2: ```suggestion
By implementing accessibility best practices, you are improving the usability of the Ceph Dashboard for blind and visually impaired users.
```

> By implementing accessibility best practices, you are also improving the usability of the site for all users.

Reviewer 3: Suggest "By following accessibility best practices we improve the usability of the site for all users"

> **Accessibility checkers:** During development, you can test the accessibility compliance of your features using one of the tools below

Reviewer 2: ```suggestion
**Accessibility checkers:** During development, you can test the accessibility compliance of your features using one of the tools below:
```

- `Accessibility insights plugin <https://accessibilityinsights.io/downloads/>`_
- `Site Improve plugin <https://www.siteimprove.com/integrations/browser-extensions/>`_
- `Axe devtools <https://www.deque.com/axe/devtools/>`_

> Testing with two or more of these tools can greatly improve the detection of possible accessibility violations

Reviewer 2: ```suggestion
Testing with two or more of these tools can greatly improve the detection of possible accessibility violations.
```

> Testing with two or more of these tools can greatly improve the detection of possible accessibility violations

Reviewer 3: s/possible//

Author: Thanks @anthonyeleven

> **Color contrast checkers:** When adding new colors, making sure they are accessible is also important, Here are some tools which can help with color contrast testing

Reviewer 2: ```suggestion
**Color contrast checkers:** When adding new colors, making sure they are accessible is also important. Here are some tools which can help with color contrast testing:
```

> **Color contrast checkers:** When adding new colors, making sure they are accessible is also important, Here are some tools which can help with color contrast testing

Reviewer 3: Suggest "When adding new colors, it is important to ensure that they are accessible.  Here are some tools that can help with contrast testing:"

- `Accessible web color-contrast checker <https://accessibleweb.com/color-contrast-checker/>`_
- `Colorsafe generator <https://colorsafe.co/>`_

> **Accessibility linters:** If you use vscode, we recommend installing the `axe accessibility linter <https://marketplace.visualstudio.com/items?itemName=deque-systems.vscode-axe-linter>`_

Reviewer 3: "VS Code"

Reviewer 3: Please s/vscode/VS Code/

Reviewer 3: Since Laura mentions the same thing I'm resolving my comment, but hers still stands.

which can help you catch and fix potential issues during development

> **Accessibility testing:** Our e2e testing suite based on cypress supports the addition of accessibility tests using `axe-core <https://github.com/dequelabs/axe-core>`_ 
> and `cypress-axe <https://github.com/component-driven/cypress-axe>`_. A custom cypress command `cy.checkAccessibility` can also be used directly. 
> This is a great way to prevent regressions on high impact components.

Reviewer 2: ```suggestion
**Accessibility testing:** Our e2e testing suite, which is based on Cypress, supports the addition of accessibility tests using `axe-core <https://github.com/dequelabs/axe-core>`_ 
and `cypress-axe <https://github.com/component-driven/cypress-axe>`_. A custom Cypress command, `cy.checkAccessibility`, can also be used directly. 
This is a great way to prevent accessibility regressions on high impact components.
```

> Test can be found under the `a11y folder <./src/pybind/mgr/dashboard/frontend/cypress/integration/a11y>`_ in the dashboard. Here is an example

Reviewer 2: ```suggestion
Tests can be found under the `a11y folder <./src/pybind/mgr/dashboard/frontend/cypress/integration/a11y>`_ in the dashboard. Here is an example:
```

.. code:: TypeScript

  describe('Navigation accessibility', { retries: 0 }, () => {
    const shared = new NavigationPageHelper();
  
    beforeEach(() => {
      cy.login();
      Cypress.Cookies.preserveOnce('token');
      shared.navigateTo();
    });
  
    it('top-nav should have no accessibility violations', () => {
      cy.injectAxe();
      cy.checkAccessibility('.cd-navbar-top');
    });
  
    it('sidebar should have no accessibility violations', () => {
      cy.injectAxe();
      cy.checkAccessibility('nav[id=sidebar]');
    });
  
});

> **Additional guidelines:** If you're unsure about which UI pattern to follow in order to implement an accessibility fix, `patternfly <https://patternfly.org/v4/>`_ guidelines can be used.

Reviewer: I'd add a short list (~5) of "things to check" (things that you've done so far in your PRs), and which are quick wins:
* Add ARIA labels and descriptions (link to ARIA labels) to actionable HTML elements,
* Don't forget to tag ARIA labels/descriptions or any user-readable text for translation (`i18-title`, `i18n-aria-label`...).
* Add ARIA roles (link to ARIA roles) to tag HTML elements that behave different from their intended behaviour (`<a>` tags behaving as `<buttons>`) or that provide extended behaviours (roles),
* Avoid poor color contrast choices (foreground-background) when styling a component (link to the afore-mentioned color contrast checker tool).
* ...

Author: Thanks @epuertat this is quite a list of easily missed items

> **Additional guidelines:** If you're unsure about which UI pattern to follow in order to implement an accessibility fix, `patternfly <https://patternfly.org/v4/>`_ guidelines can be used.

Reviewer: Please, use specific links:
```suggestion
**Additional guidelines:** If you're unsure about which UI pattern to follow in order to implement an accessibility fix, `patternfly <https://www.patternfly.org/v4/accessibility/accessibility-fundamentals>`_ guidelines can be used.
```

Default notification should have `text-info` color. Success notification should
have `text-success` color. Failure notification should have `text-danger` color.

:ref:`Accessibility <accessibility>`
....................................

> 

Reviewer 2: @nSedrickm this needs to be removed since you put "Accessibility" in a new section.

For example - ``throw new DashboardNotFoundError()``.

> Internationalization (i18n)

Reviewer: Thank you @nSedrickm !

  <!-- recommended -->
  <span i18n>Profile <b>foo</b> will be removed.</span>

.. _accessibility:

Accessibility
-------------

Many parts of the Ceph Dashboard are modeled on `Web Content Accessibility Guidelines (WCAG) 2.1 <https://www.w3.org/TR/WCAG21/>`_  level A accessibility conformance guidelines. 
By implementing accessibility best practices, you are improving the usability of the Ceph Dashboard for blind and visually impaired users.

Accessibility checkers
~~~~~~~~~~~~~~~~~~~~~~

During development, you can test the accessibility compliance of your features using one of the tools below:

- `Accessibility insights plugin <https://accessibilityinsights.io/downloads/>`_
- `Site Improve plugin <https://www.siteimprove.com/integrations/browser-extensions/>`_
- `Axe devtools <https://www.deque.com/axe/devtools/>`_

Testing with two or more of these tools can greatly improve the detection of accessibility violations.

.. _color-contrast-checkers:

Color contrast checkers
~~~~~~~~~~~~~~~~~~~~~~~

> When adding new colors, making sure they are accessible is also important. Here are some tools which can help with color contrast testing

Reviewer 2: ```suggestion
When adding new colors, making sure they are accessible is also important. Here are some tools which can help with color contrast testing:
```

- `Accessible web color-contrast checker <https://accessibleweb.com/color-contrast-checker/>`_
- `Colorsafe generator <https://colorsafe.co/>`_

Accessibility linters
~~~~~~~~~~~~~~~~~~~~~

> If you use VSCode, you may install the `axe accessibility linter <https://marketplace.visualstudio.com/items?itemName=deque-systems.vscode-axe-linter>`_

Reviewer 2: ```suggestion
If you use VSCode, you may install the `axe accessibility linter <https://marketplace.visualstudio.com/items?itemName=deque-systems.vscode-axe-linter>`_ ,
```

which can help you catch and fix potential issues during development.

Accessibility testing
~~~~~~~~~~~~~~~~~~~~~

> Our e2e testing suite which is based on Cypress supports the addition of accessibility tests using `axe-core <https://github.com/dequelabs/axe-core>`_ 

Reviewer 2: ```suggestion
Our e2e testing suite, which is based on Cypress, supports the addition of accessibility tests using `axe-core <https://github.com/dequelabs/axe-core>`_ 
```

> and `cypress-axe <https://github.com/component-driven/cypress-axe>`_. A custom Cypress command `cy.checkAccessibility`, can also be used directly. 

Reviewer 2: ```suggestion
and `cypress-axe <https://github.com/component-driven/cypress-axe>`_. A custom Cypress command, `cy.checkAccessibility`, can also be used directly. 
```

This is a great way to prevent accessibility regressions on high impact components.

Tests can be found under the `a11y folder <./src/pybind/mgr/dashboard/frontend/cypress/integration/a11y>`_ in the dashboard. Here is an example:

.. code:: TypeScript

  describe('Navigation accessibility', { retries: 0 }, () => {
    const shared = new NavigationPageHelper();
  
    beforeEach(() => {
      cy.login();
      Cypress.Cookies.preserveOnce('token');
      shared.navigateTo();
    });
  
    it('top-nav should have no accessibility violations', () => {
      cy.injectAxe();
      cy.checkAccessibility('.cd-navbar-top');
    });
  
    it('sidebar should have no accessibility violations', () => {
      cy.injectAxe();
      cy.checkAccessibility('nav[id=sidebar]');
    });
  
  });

Additional guidelines
~~~~~~~~~~~~~~~~~~~~~

If you're unsure about which UI pattern to follow in order to implement an accessibility fix, `patternfly <https://www.patternfly.org/v4/accessibility/accessibility-fundamentals>`_ guidelines can be used.

> 

Reviewer 2: ```suggestion

A few things you should check before introducing a new code change include:
```
And make this list numbered.

> - Add `ARIA labels and descriptions <https://www.w3.org/TR/wai-aria/>`_ to actionable HTML elements,
> - Don't forget to tag ARIA labels/descriptions or any user-readable text for translation (i18n-title, i18n-aria-label...).
> - Add `ARIA roles <https://www.w3.org/TR/wai-aria/#usage_intro>`_ to tag HTML elements that behave different from their intended behaviour (<a> tags behaving as <buttons>) or that provide extended behaviours (roles),
> - Avoid poor `color contrast choices <https://www.w3.org/TR/WCAG21/#contrast-minimum>`_ (foreground-background) when styling a component. Here are some :ref:`tools <color-contrast-checkers>` you can use.
> - When testing menus or dropdowns, be sure to scan them in both opened and closed states. Sometimes issues are hidden when menus are closed
> 

Reviewer 2: To keep the punctuation here consistent:
```suggestion
- Add `ARIA labels and descriptions <https://www.w3.org/TR/wai-aria/>`_ to actionable HTML elements.
- Don't forget to tag ARIA labels/descriptions or any user-readable text for translation (i18n-title, i18n-aria-label...).
- Add `ARIA roles <https://www.w3.org/TR/wai-aria/#usage_intro>`_ to tag HTML elements that behave different from their intended behaviour (<a> tags behaving as <buttons>) or that provide extended behaviours (roles).
- Avoid poor `color contrast choices <https://www.w3.org/TR/WCAG21/#contrast-minimum>`_ (foreground-background) when styling a component. Here are some :ref:`tools <color-contrast-checkers>` you can use.
- When testing menus or dropdowns, be sure to scan them in both opened and closed states. Sometimes issues are hidden when menus are closed.
```

      cy.checkAccessibility('nav[id=sidebar]');
    });
  
  });

Additional guidelines
~~~~~~~~~~~~~~~~~~~~~

If you're unsure about which UI pattern to follow in order to implement an accessibility fix, `patternfly <https://www.patternfly.org/v4/accessibility/accessibility-fundamentals>`_ guidelines can be used.

A few things you should check before introducing a new code change include:

- Add `ARIA labels and descriptions <https://www.w3.org/TR/wai-aria/>`_ to actionable HTML elements.
- Don't forget to tag ARIA labels/descriptions or any user-readable text for translation (i18n-title, i18n-aria-label...).
- Add `ARIA roles <https://www.w3.org/TR/wai-aria/#usage_intro>`_ to tag HTML elements that behave different from their intended behaviour (<a> tags behaving as <buttons>) or that provide extended behaviours (roles).
- Avoid poor `color contrast choices <https://www.w3.org/TR/WCAG21/#contrast-minimum>`_ (foreground-background) when styling a component. Here are some :ref:`tools <color-contrast-checkers>` you can use.
- When testing menus or dropdowns, be sure to scan them in both opened and closed states. Sometimes issues are hidden when menus are closed.

> 

Reviewer: Now I see this, and given the informational overload we're exposed to every day, why not move this to the top of the section (sort of "TL;DR" summary).

And I'd say "follow this 5 rules" and use a numbered list to emphasize the 5-commandments nature of this.

  <!-- recommended -->
  <span i18n>Profile <b>foo</b> will be removed.</span>

.. _accessibility:

Accessibility
-------------

Many parts of the Ceph Dashboard are modeled on `Web Content Accessibility Guidelines (WCAG) 2.1 <https://www.w3.org/TR/WCAG21/>`_  level A accessibility conformance guidelines. 
By implementing accessibility best practices, you are improving the usability of the Ceph Dashboard for blind and visually impaired users.

Summary
~~~~~~~

A few things you should check before introducing a new code change include:

1) Add `ARIA labels and descriptions <https://www.w3.org/TR/wai-aria/>`_ to actionable HTML elements.
2) Don't forget to tag ARIA labels/descriptions or any user-readable text for translation (i18n-title, i18n-aria-label...).
3) Add `ARIA roles <https://www.w3.org/TR/wai-aria/#usage_intro>`_ to tag HTML elements that behave different from their intended behaviour (<a> tags behaving as <buttons>) or that provide extended behaviours (roles).
4) Avoid poor `color contrast choices <https://www.w3.org/TR/WCAG21/#contrast-minimum>`_ (foreground-background) when styling a component. Here are some :ref:`tools <color-contrast-checkers>` you can use.

> 5) When testing menus or dropdowns, be sure to scan them in both opened and closed states. Sometimes issues are hidden when menus are closed.

Reviewer 2: ```suggestion
5) When testing menus or dropdowns, be sure to scan them with an accessibility checker in both opened and closed states. Sometimes issues are hidden when menus are closed.
```
Then link the words "accessibility checker" to the section on it you have below.
